src/lib/stable/canister_methods/index.ts

```diff
@@ -62,104 +80,151 @@ export function decoratorArgumentsHandler<This, Args extends any[], Return>(
     }
 }
 
-function decoratorImplementation<This, Args extends any[], Return>(
+function isDecoratorOverloadedWithoutParams<This, Args extends any[], Return>(
+    param1?: MethodType<This, Args, Return> | IDL.Type[],
+    param2?:
+        | ClassMethodDecoratorContext<This, MethodType<This, Args, Return>>
+        | IDL.Type
+): boolean {
+    return (
+        typeof param1 === 'function' &&
+        param2 !== undefined &&
+        'kind' in param2 &&
+        param2.kind === 'method' &&
+        param2.metadata !== undefined &&
+        param2.name !== undefined
+    );
+}
+
+function decoratorImplementation<
+    This extends ExportedCanisterClass,
+    Args extends any[],
+    Return
+>(
     canisterMethodMode: CanisterMethodMode,
     originalMethod: MethodType<This, Args, Return>,
-    context: ClassMethodDecoratorContext,
+    context: ClassMethodDecoratorContext<This, MethodType<This, Args, Return>>,
     paramIdlTypes?: IDL.Type[],
     returnIdlType?: IDL.Type,
     options?: { composite?: boolean; manual?: boolean }
 ): MethodType<This, Args, Return> {
-    const name = context.name as string;
-
-    const index = globalThis._azleCanisterMethodsIndex++;
-    const indexString = index.toString();
-
-    if (canisterMethodMode === 'query') {
-        globalThis._azleMethodMeta.queries?.push({
-            name,
-            index,
-            composite: options?.composite ?? false
-        });
-
-        globalThis._azleCanisterMethodIdlTypes[name] = IDL.Func(
-            paramIdlTypes ?? [],
-            returnIdlType === undefined ? [] : [returnIdlType],
-            ['query']
-        );
-    }
+    context.addInitializer(function () {
```

Comment:
If I understand correctly, for this file at least this is the biggest change is that everything happens inside this initializer right? Then instead of modifying the global stuff we modify the class stuff and then just add this stuff all in one go to the global stuff? Can you go over with me one more time what exactly the add Initializer function does?